fs/nvs: operate on filled up storage bugfixes

nvlsianpu opened the pull request:
- test case for filled up fiesystem
- fix for bug: unable to delete anything from filled up system
- fix for bug: initialization can hang/assert on filled up sector
- skip deletion of non-existing entry
- test case for deletion  entry behaviour

patch is backward compatible - older storage instance in a flash will work with it.

nvlsianpu committed:
fs/nvs: fix initialization hang if sector is full

During initialization NVS analyzes open sector for
find write addresses. It was possible to infinite loop
in case when there was no space for any new ATE in the sector.

This patch introduce check for space available in the sector which
fix the issue.

fixes #17891

Signed-off-by: Andrzej Puzdrowski <andrzej.puzdrowski@nordicsemi.no>

nvlsianpu committed:
tests/subsys/fs/nvs: init on full storage test

Added test for check wheter initialization works well when nvs
storage is full filled. Test also check weheter can delet any
entry from full filled NVS.

Signed-off-by: Andrzej Puzdrowski <andrzej.puzdrowski@nordicsemi.no>

nvlsianpu committed:
fs/nvs: fix delete ate accommodation

NVS always leave space for deletion ate in a full sector.
Even that it was not possible to write delete nvs_ate to the
full sector. Because of that it was possible to fill up NVS,
and be able to delete nothing.

This patch introduce recognition of case
the delete ate is writing, and allow to write such nvs_ate to
opened full sector.

Signed-off-by: Andrzej Puzdrowski <andrzej.puzdrowski@nordicsemi.no>

nvlsianpu committed:
fs/nvs: skip deletion on non existing entry

No need to write deletion entry for non-existing entry.
Furthermore such a write to filled up storage will make another
writes impossible as it occupied delete nvs_ate space, but it
does not remove anything.

Signed-off-by: Andrzej Puzdrowski <andrzej.puzdrowski@nordicsemi.no>

nvlsianpu commented:
cc @kapi-no Please review.

Laczen commented on subsys/fs/nvs/nvs.c:
```diff
@@ -526,6 +526,19 @@ static int nvs_gc(struct nvs_fs *fs)
 	return 0;
 }
 
+static int nvs_mem_cmp_const(const u8_t *buf, u8_t const_val, size_t len)
```
This routine is already available (nvs_ate_cmp_cnst), no need to add a new one.

nvlsianpu replied:
right, I didn't notice

Laczen commented on subsys/fs/nvs/nvs.c:
```diff
@@ -574,17 +587,42 @@ static int nvs_startup(struct nvs_fs *fs)
 		addr -= ate_size;
-		rc = nvs_flash_cmp_const(fs, addr, 0xff,
-					  sizeof(struct nvs_ate));
+
```
Why all this complicated stuff, just check if the address reached a sector start. If this is the case return an error.

kapi-no replied:
I would also say that it's better to fail hard when detecting any inconsistency in the FS structure (without any auto-recovery attempts).

nvlsianpu replied:
There is no auto recovery here. the latest ATE might be just after the latest DATA. This if statement will detect that and will break the while much earlier than `if` in line 595 (than ate scanning come to sector begin). Line 595 is the last resore - it serves damaged FS/not-NVS case, while this one serves just filled up sector.

Laczen commented on subsys/fs/nvs/nvs.c:
```diff
@@ -737,7 +737,7 @@ ssize_t nvs_write(struct nvs_fs *fs, u16_t id, const void *data, size_t len)
 	struct nvs_ate wlk_ate;
 	u32_t wlk_addr, rd_addr;
-	u16_t sector_freespace;
+	s32_t sector_freespace;
```
No need to change this, a sector can maximum be 32kB.

kapi-no replied:
why? address space for each sector seems to be in the following range:
0x0000 - 0xFFFF

nvlsianpu replied:
If sector is filled up (delete ATE is written at the last possible position) free-space is below zero.

Laczen replied:
@nvlsianpu, this should never be the case, it would appear when a delete is written to a location that would not allow space for an empty ate to be added. This cannot be allowed.

Laczen commented on subsys/fs/nvs/nvs.c:
```diff
@@ -805,11 +805,12 @@ ssize_t nvs_write(struct nvs_fs *fs, u16_t id, const void *data, size_t len)
-
+		if ((sector_freespace >= (s32_t)(data_size + ate_size)) ||
+		    /* accommodate the delete ate */
+		    (len == 0 && sector_freespace >= 0)) {
```
I don't think what is done here is correct, when data is written 2 ate's should be available, one for a possible delete and one for an empty ate. This empty ate is needed whenever a write to the sector end would fail, in that case the empty ate becomes the only way to determine the last written ate.

Laczen replied:
A simple fix:
```
if (len == 0) {
    sector_freespace += ate_size;
}
if (sector_freespace >= data_size + 2 * ate_size) {
```

kapi-no replied:
So there was a mistake with previous implementation (because it didn't take empty ATE into account) ?

```if (sector_freespace >= data_size + ate_size) {```

Laczen replied:
@kapi-no, yes and no: this is a border case where all entries would fit exactly, in most cases there will be sufficient place to add the delete. It is also questionable if this functionality is needed, with the correction applied there always is the possibility (even in the border case) to delete entries and free up space in the fs, in the previous implementation the fs would be full and you could only erase it to free up space. I am unsure what is best, if the delete functionality is always available and used it might go trough the flash lifetime very fast.

nvlsianpu replied:
I can said that my current fix is proper. Previously case of writing entry with data was properly served as well. Problem touched only delete ATE.

nvlsianpu replied:
> I don't think what is done here is correct, when data is written 2 ate's should be available, one for a possible delete and one for an empty ate. This empty ate is needed whenever a write to the sector end would fail, in that case the empty ate becomes the only way to determine the last written ate.

This is new idea that one empty ATE should be keep always.
Maximum filling for the sector will then looks like:
```
|n* data|empty-ATE|delete-ATE|k * ATE|
```

nvlsianpu replied:
> This is new idea that one empty ATE should be keep always.

Or just keeping place for the last delete ATE was not asserted - then it might been original idea.

nvlsianpu replied:
I would update design documentation (i think we don't have a such) about that.

Laczen replied:
@nvlsianpu, I don't quite get what your decision would be:
a. Always keep space for a delete: then you would need to check that there are 2 ate's available when a data write is performed (the first part of the if clause).

or

b. Don't keep space for a delete, then you can keep the old code.

nvlsianpu replied:
@Laczen For sure I always want to be able to delete something, even if nvs is full. What I done was with assumption that the one empty ate is unneeded (I think that always better to have this 8 byte). Honestly I was thinking that empty ate was reserved to be used for the last delete, and last delete was not possible by a bug. So I'm investigating the code whether I can really get rid of empty ATE requirement (even that I worked hard trying fix that). If I don't find anything better I will go your way.

Laczen replied:
@nvlsianpu, after going through the code again I realized that I was wrong regarding the empty ate, this is only needed during the startup (to stop addr -= ate_size). Your method should work. I would apply a small change and use a required_space calculation instead of free space (line 808):
```
	u16_t required_space;
		...

		required_space = 0;
		if (data_size) {
			/* Leave space for delete ate */
			required_space += data_size + ate_size;
		}
		if (fs->ate_wra >= fs->data_wra + required_space) {

			rc = nvs_flash_wrt_entry(fs, id, data, len);
			if (rc) {
				goto end;
			}
			break;
		}
	...
```

nvlsianpu replied:
I Applied this suggestion before the while loop.

kapi-no commented on subsys/fs/nvs/nvs.c:
```diff
@@ -574,17 +587,42 @@ static int nvs_startup(struct nvs_fs *fs)
 	 */
 	while (1) {
+		if ((addr & ADDR_OFFS_MASK) < ate_size) {
+			/* return -EDEADLK; */
```
Why did you put this comment here if we are only breaking out of the loop and not returning an error?

nvlsianpu replied:
Initially I had considered this as an error in storage, but then I found out that this can be proper state as well. This commented line will be removed.

Laczen replied:
@nvlsianpu, it cannot be a proper state. There should always be one empty ate in a sector. Otherwise there would be no fallback option when the sector end ate write would fail. NVS considers each ate to be valid (but it will not use the ate when the crc8 is invalid). If there is no empty ate you will always reach the 0 address.

Laczen replied:
@nvlsianpu, @kapi-no, something that is wrong with nvs is that a fallback method is needed for the end sector ate. This could be avoided by not writing a sector end ate but by writing the needed information to the start of the next sector, in that case it could be checked at startup (and/or after the ate write) if the ate is correct and if needed erased and rewritten. Although it is not so straight forward (a combination of the ate write address, crc8 and data offset should be used to determine the last valid ate), it is possible.

kapi-no commented on tests/subsys/fs/nvs/src/main.c:
```diff
@@ -484,6 +484,45 @@ void test_nvs_corrupted_sector_close_operation(void)
+	len = nvs_write(&fs, 1, &filling_id, sizeof(filling_id));
+	zassert_true(len == sizeof(filling_id), "nvs_write failed: %d", len);
+}
+
```
Consider adding a test to check if delete ATE can be written to the NVS with an ID that doesn't match any other IDs.

nvlsianpu replied:
good point

nvlsianpu replied:
test of delete behavior added

Laczen commented:
I have taken another look at the startup routine and this is my proposal:
```
static int nvs_startup(struct nvs_fs *fs)
{
	int rc;
	struct nvs_ate last_ate;
	size_t ate_size, empty_len;
	/* Initialize addr to 0 for the case fs->sector_count == 0. This
	 * should never happen as this is verified in nvs_init() but both
	 * Coverity and GCC believe the contrary.
	 */
	u32_t addr = 0U;
	u16_t i, closed_sectors = 0;

	k_mutex_lock(&fs->nvs_lock, K_FOREVER);

	ate_size = nvs_al_size(fs, sizeof(struct nvs_ate));
	/* step through the sectors to find a open sector following
	 * a closed sector, this is where NVS can to write.
	 */
	for (i = 0; i < fs->sector_count; i++) {
		addr = (i << ADDR_SECT_SHIFT) + fs->sector_size - ate_size;
		rc = nvs_flash_cmp_const(fs, addr, 0xff,
					  sizeof(struct nvs_ate));
		if (rc) {
			/* closed sector */
			closed_sectors++;
			nvs_sector_advance(fs, &addr);
			rc = nvs_flash_cmp_const(fs, addr, 0xff,
						  sizeof(struct nvs_ate));
			if (!rc) {
				/* open sector */
				break;
			}
		}
	}
	/* all sectors are closed, this is not a nvs fs */
	if (closed_sectors == fs->sector_count) {
		return -EDEADLK;
	}

	if (i == fs->sector_count) {
		/* none of the sectors where closed, in most cases we can set
		 * the address to the first sector, except when there are only
		 * two sectors. Then we can only set it to the first sector if
		 * the last sector contains no ate's. So we check this first
		 */
		rc = nvs_flash_cmp_const(fs, addr - ate_size, 0xff,
				  sizeof(struct nvs_ate));
		if (!rc) {
			/* empty ate */
			nvs_sector_advance(fs, &addr);
		}
	}

	/* addr contains address of the last ate in the most recent sector,
	 * update the ate_wra and data_wra by stepping through the sector
	 */
	addr -= ate_size;
	fs->ate_wra = addr - ate_size;
	fs->data_wra = addr & ADDR_SECT_MASK;
	
	while (fs->ate_wra > fs->data_wra) {
		rc = nvs_flash_ate_rd(fs, fs->ate_wra, &last_ate);
		if (rc) {
			goto end;
		}

		rc = nvs_ate_cmp_const(&last_ate, 0xff);
		if (!rc) {
			/* found ff empty location */
			break;
		}

		if (!nvs_ate_crc8_check(&last_ate)) {
			/* crc8 is ok, complete write of ate was performed */
			fs->data_wra += last_ate.offset;
			fs->data_wra += nvs_al_size(fs, last_ate.len);
		}

		fs->ate_wra -= ate_size;
	}

	/* if fs->ate_wra is pointing to the start of the sector there was no
	 * valid data in the sector, this is not a nvs fs
	 */
	if ((fs->ate_wra & ADDR_SECT_MASK) == 0) {
		return -EDEADLK;
	}

	/* possible data write after last ate write, update data_wra */
	while (fs->ate_wra > fs->data_wra) {
		
		empty_len = fs->ate_wra - fs->data_wra;
		
		rc = nvs_flash_cmp_const(fs, fs->data_wra, 0xff, empty_len);
		if (rc < 0) {
			goto end;
		}
		if (!rc) {
			break;
		}
		fs->data_wra += fs->write_block_size;
	}

	/* if the sector after the write sector is not empty gc was interrupted
	 * we need to restart gc, first erase the sector before restarting gc
	 * otherwise the data may not fit into the sector.
	 */
	addr = fs->ate_wra & ADDR_SECT_MASK;
	nvs_sector_advance(fs, &addr);
	rc = nvs_flash_cmp_const(fs, addr, 0xff, fs->sector_size);
	if (rc < 0) {
		goto end;
	}
	if (rc) {
		/* the sector after fs->ate_wrt is not empty */
		rc = nvs_flash_erase_sector(fs, fs->ate_wra);
		if (rc) {
			goto end;
		}
		fs->ate_wra &= ADDR_SECT_MASK;
		fs->ate_wra += (fs->sector_size - 2 * ate_size);
		fs->data_wra = (fs->ate_wra & ADDR_SECT_MASK);
		rc = nvs_gc(fs);
		if (rc) {
			goto end;
		}
	}

end:
	k_mutex_unlock(&fs->nvs_lock);
	return rc;
}
```

nvlsianpu commented:
@Laczen 
```
/* if fs->ate_wra is pointing to the start of the sector there was no
	 * valid data in the sector, this is not a nvs fs
	 */
```

Considering sector which contains only delete ATEs up to full - It is possible. Am I right?

Laczen commented:
> @Laczen
> 
> ```
> /* if fs->ate_wra is pointing to the start of the sector there was no
> 	 * valid data in the sector, this is not a nvs fs
> 	 */
> ```
> 
> Considering sector which contains only delete ATEs up to full - It is possible. Am I right?

You are right, this is possible. The nvs_crc check could be used to set a bool valid=yes and check the bool to error out of startup.

nvlsianpu commented:
^^ OK will apply your suggestions.

nvlsianpu committed:
tests/subsys/fs/nvs: entry delete test

Added test for check behavior on deleting existent entry,
nonexistent entry and already deleted entry.

Signed-off-by: Andrzej Puzdrowski <andrzej.puzdrowski@nordicsemi.no>

nvlsianpu commented:
@Laczen, @kapi-no updated

nvlsianpu commented on subsys/fs/nvs/nvs.c:
```diff
@@ -737,7 +737,7 @@ ssize_t nvs_write(struct nvs_fs *fs, u16_t id, const void *data, size_t len)
 	u32_t wlk_addr, rd_addr;
-	u16_t sector_freespace;
+	u16_t required_space;
 
```
Value for this variable is assigned in any flow, so shouldn't be an issue.
Oh, missed U qualifier, so will apply suggestion anyway.

eHammarstrom commented:
@nvlsianpu does this introduce a bug where if the last open sector is closed before a gc is executed and the device reboots we will report EDEADLK when we could have restarted the gc?

Laczen commented:
> @nvlsianpu does this introduce a bug where if the last open sector is closed before a gc is executed and the device reboots we will report EDEADLK when we could have restarted the gc?

This shouldn't be the case, when a sector is closed the next sector is already by the previous gc.

eHammarstrom commented:
@Laczen When a sector is full we will close it and start GC of the next sector before we erase it for further usage.
If we crash after the close but before the GC of the next sector, on next startup we will believe that all sectors closed with the addition of this patch because we have not _definitively_ marked that the next sector should be GC'd, and even if we did this patch does not consider that entry when it checks if all sectors are used.

eHammarstrom commented:
@Laczen I think the issue is if we crash during GC before the next erase.

```
---
Active
---
Erased
---
Next GC target
---
```

If `Active` becomes full, we start GC into `Erased`, if we crash during the GC then by the check added in this patch (checking if last ATE is 0xFF, but it's not anymore since we started GC) we will consider all sectors closed.

Laczen commented:
@eHammarstrom for a better understanding from my side: have you experienced this on the latest version or is this on an old version ? On what device and what is the configuration ?

eHammarstrom commented:
@Laczen We're using Zephyr 3.3 currently.

The partition is located in an external QSPI-connected flash on an nRF5340 and the partition size is 64kB and each sector is 4kB.

Laczen commented:
OK, I have been giving this some more thinking and I can't see it happening:

The sector closing is not done by gc but by the writing of data to a sector. Suppose it fails during gc then the sector that is used to write data to is not closed (gc doesn't do that), when gc is done the next sector will be erased. Even in the case it would fail during the erase the sector that is used to write data is not closed. Suppose it fails just after the closing of a sector, in that case the next sector would have been erased already.

Are you sure there is no overlap or something else writing to the area ? Or has there been a changed in the sector definition ?

nvlsianpu commented:
Are you sure that this is not caused by garbage in your memory? NVS should be established on already erased memory.
So need to ensure that NVS memory is erase while you flashing your device.
